src/memory-analyzer/gdb_api.h

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************\
+
+Module: GDB Machine Interface API
+
+Author: Malte Mues <mail.mues@gmail.com>
+        Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Low-level interface to gdb
+///
+/// This provides an API to run a program under gdb up to some
+/// breakpoint, and then query the values of expressions. It uses the
+/// gdb machine interface (see section "The GDB/MI Interface" in the
+/// gdb manual to communicate with gdb.
+
+// clang-format off
+#if defined(__linux__) || \
+    defined(__FreeBSD_kernel__) || \
+    defined(__GNU__) || \
+    defined(__unix__) || \
+    defined(__CYGWIN__) || \
+    defined(__MACH__)
+// clang-format on
+
+#ifndef CPROVER_MEMORY_ANALYZER_GDB_API_H
+#define CPROVER_MEMORY_ANALYZER_GDB_API_H
+#include <unistd.h>
+
+#include <exception>
+
+#include <util/exception_utils.h>
+
+class gdb_apit
+{
+public:
+  explicit gdb_apit(const char *binary, const bool log = false);
+
+  void create_gdb_process();
+  void terminate_gdb_process();
+
+  bool run_gdb_to_breakpoint(const std::string &breakpoint);
+  void run_gdb_from_core(const std::string &corefile);
+
+  std::string get_value(const std::string &expr);
+  std::string get_memory(const std::string &expr);
+
+  const std::vector<std::string> &get_command_log();
+
+protected:
+  const char *binary;
+
+  FILE *input_stream;
+  FILE *output_stream;
+
+  const bool log;
+  std::vector<std::string> command_log;
```

Comment:
Why not, it's just the with `push_front` the commands will be stored in reversed order.

Reply:
@tautschnig Could you explain why a `forward_list` here is better? Thanks!

Reply:
I was thinking that `command_log` would be an append-only data structure. With a `std::vector`, each append operation incurs considerable cost. But likely there is no measurable difference, because this `command_log` will never be particularly big.